### ISSUE TITLE
Help setting up latching solenoid/relay

### ISSUE THREAD

**Issue by pyrosmiley** [OttoWinter/esphomeyaml#104]
Hi Otto, thanks for dealing with my dumb issues as they come up! I really have scoured the docs for this one and it _seems_ like I should be able to do it, but I can't _quite_ get it there. 

I need to control a latching solenoid valve, with two separate GPIO pins going to an h-bridge. Right now I have two `switch` instances set up, one each for `set` and `reset` functions, and two `template switch`  for `on` and `off` that pulses the proper GPIO output, and that's what's exposed to Home Assistant. For simply turning on and off a valve, it works. This is all based on the [Template Cover](https://esphomelib.com/esphomeyaml/components/cover/template.html) example I found in the docs. Unfortunately, there's no sensor to help determine the state. 

Normally I'd just let that be that and move on with life, except for the fact that I also need to use Deep Sleep, and I have to be able to have a value to store in nonvolatile memory (or, ideally, the [RTC Memory](https://www.youtube.com/watch?v=r-hEOL007nw)!). I haven't gotten to the stage of figuring out how I'm gonna do that yet but I know it should be do-able because you already have it implemented with the `light` component so I'll probably just poke around in the C code for those. In an effort to be other-user-friendly, I also need to be able to toggle the state of the valve with a single button for each, but that'll be simple once I have a unified switch to actually toggle.

My other plan was to try and just set it up in Home Assistant, which isn't ideal, but might work since it stays up while the ESPs are sleeping. Unfortunately, ESPHY is smart enough to turn off the all the outputs when it wakes up. Not so good for 3 minute deep sleep when I wanna water the garden for 30 minutes....

In the meantime, I've been trying and failing to set one up with a `template binary sensor` and an API call to toggle the state as part of the `set` and `reset` actions. I thought for sure this would be the method that works, but alas, I can't seem to write it such that it will have an initial value without referencing something else, and there's nothing to reference. 
tl;dr
- Separate pulsed GPIO pins for `set` and `reset` ✔️
- Templated `state` value based not on the physical outputs, but the `on` and `off` actions of unified template switch 😕
 - Save state to nonvolatile memory for persistence across deep sleeps 🤷‍♂️

I really hope I'm not missing something, but I asked on the Discord to no avail, and my head is just spinning in circles to make this one work.

**Comment by OttoWinter** [comment]
> Templated state value based not on the physical outputs, but the on and off actions of unified template switch 😕

I don't understand, why not just set `optimistic` to True. Then the template switch is in charge of storing the state.

> Save state to nonvolatile memory for persistence across deep sleeps 🤷‍♂️

Problem with that is that on the ESP8266 I've not yet been able to get RTC to work reliably (thus also the lack of automatic restore of switches). The ESP32 *does* have awesome non-volatile storage libraries that make it really easy to store values across boots. But with the ESP8266 you only really have the RTC memory and using that quickly leads to some very weird OTA errors.

**Comment by pyrosmiley** [comment]
Ah okay, interesting!! I didn't realize RTC memory messes with OTA. Definitely do _not_ want to break that!

I also did not quite realize that that's how `optimistic` works.... That's my bad -- I guess it just didn't stick with me somehow. Thanks!
